src/Bicep.Core/PrettyPrint/PrettyPrinter.cs

```diff
@@ -0,0 +1,67 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using Bicep.Core.Extensions;
+using Bicep.Core.Parser;
+using Bicep.Core.PrettyPrint.Options;
+using Bicep.Core.Syntax;
+
+namespace Bicep.Core.PrettyPrint
+{
+    public static class PrettyPrinter
+    {
+        private static readonly IReadOnlyDictionary<int, string> IndentsBySize = Enumerable.Range(1, 8)
+            .ToDictionary(size => size, size => new string(' ', size));
+
+        public static string? PrintProgram(ProgramSyntax programSyntax, PrettyPrintOptions options)

```

Comment:
>PrintProgram [](start = 30, length = 12)

I haven't read the rest of the code yet, but I'm wondering something... We have some cases where we need to produce dynamically generated multi-line snippet completions. I want to make sure we adhere to the user's formatting settings (spaces, indent size, etc.) when we do so the completion doesn't produce code that looks out of place. How hard would be to modify this in the future to support formatting a subtree of the AST? (Out of scope of this PR, of course.)

Reply:
We can change the method to take a SyntaxBase, but the tricky thing is how we can fix the indent. There are other many other edge cases to handle. Waldler's paper doesn't mention anything about range formatting, but Prettier implements it. My idea is to find the top level statements that overlap the range in a formatting request and then format the statements, which should work but it's not optimal. I'm sure prettier does something more advanced, and I'll take a look at their source code to figure out.

Reply:
Yeah for that specific scenario, committing the completion would reformat stuff around it, which would be kind of weird. I haven't seen how auto formatting normally works in VS Code. In VS in C# auto formatting only triggers when you type `;` or close a block - they don't seem to do it when you commit a completion. I'm mostly worried about user experience and less worried about efficiency. Although if we got both, that works too.